README.md
# Rust VEML6070 UVA Light Sensor Driver

[![crates.io](https://img.shields.io/crates/v/veml6070.svg)](https://crates.io/crates/veml6070)
[![Docs](https://docs.rs/veml6070/badge.svg)](https://docs.rs/veml6070)
[![Build Status](https://github.com/eldruin/veml6070-rs/workflows/Build/badge.svg)](https://github.com/eldruin/veml6070-rs/actions?query=workflow%3ABuild)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/veml6070-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/veml6070-rs?branch=master)

This is a platform agnostic Rust driver for the VEML6070 UVA light sensor,
based on the [`embedded-hal`](https://github.com/rust-embedded/embedded-hal) traits.

This driver allows you to:
- Enable/disable the sensor. See: `enable()`.
- Read the UV measurement. See: `read_uv()`.
- Set the integration time. See: `set_integration_time()`.
- Enable/disable ACK signal. See: `enable_ack()`.
- Set ACK threshold value. See: `set_ack_threshold()`.

## The device
VEML6070 is an advanced ultraviolet (UV) light sensor with I2C protocol
interface and designed by the CMOS process.
It is easily operated via a simple I2C command. The active acknowledge
(ACK) feature with threshold windows setting allows the UV sensor to
send out a UVI alert message. Under a strong solar UVI condition, the
smart ACK signal can be easily implemented by the software programming.
VEML6070 incorporates a photodiode, amplifiers, and analog / digital
circuits into a single chip. VEML6070's adoption of Filtron TM UV
technology provides the best spectral sensitivity to cover UV spectrum
sensing. It has an excellent temperature compensation and a robust refresh
rate setting that does not use an external RC low pass filter.
VEML6070 has linear sensitivity to solar UV light and is easily adjusted
by an external resistor.
Software shutdown mode is provided, which reduces power consumption to
be less than 1 μA. VEML6070's operating voltage ranges from 2.7 V to 5.5 V.

Datasheet:
- [VEML6070](https://www.vishay.com/docs/84277/veml6070.pdf)

Application note:
- [VEML6070 AN](https://www.vishay.com/docs/84310/designingveml6070.pdf)

## Usage example

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the device.

Please find additional examples in this repository: [driver-examples]

[driver-examples]: https://github.com/eldruin/driver-examples

```rust
use linux_embedded_hal::I2cdev;
use veml6070::Veml6070;

fn main() {
    let dev = I2cdev::new("/dev/i2c-1").unwrap();
    let mut uv_light_sensor = Veml6070::new(dev);
    // initialization step is necessary
    uv_light_sensor.init().unwrap();
    uv_light_sensor.enable().unwrap();
    let reading = uv_light_sensor.read_uv().unwrap();
    println!("UV reading: {}", reading);
}
```

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/eldruin/veml6070-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed
- [breaking-change] Renamed VEML6070 -> Veml6070 to comply with Rust naming conventions.
- [breaking-change] Remove deriving `Default` for `Veml6070`.
- Removed `bit_field` dependency
- Make types `AckThreshold` and `IntegrationTime` `Copy` and `PartialEq`.

## 0.1.0 - 2018-10-12

This is the initial release to crates.io of the feature-complete driver. There
may be some API changes in the future, in case I decide that something can be
further improved. All changes will be documented in this CHANGELOG.

[Unreleased]: https://github.com/eldruin/veml6070-rs/compare/v0.1.0...HEAD



tests/tests.rs
use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTrans};
use veml6070::{AckThreshold, IntegrationTime, Veml6070};

struct Address;

impl Address {
    const ARA: u8 = 0x0C;
    const COMMAND: u8 = 0x38;
    const DATA_MSB: u8 = 0x39;
    const DATA_LSB: u8 = 0x38;
}

struct BitFlags;

impl BitFlags {
    const SHUTDOWN: u8 = 0b0000_0001;
    const IT0: u8 = 0b0000_0100;
    const IT1: u8 = 0b0000_1000;
    const ACK_THD: u8 = 0b0001_0000;
    const ACK: u8 = 0b0010_0000;
}

const DEFAULT_CMD: u8 = 0x02;

fn new(transactions: &[I2cTrans]) -> Veml6070<I2cMock> {
    Veml6070::new(I2cMock::new(transactions))
}

fn destroy(dev: Veml6070<I2cMock>) {
    dev.destroy().done();
}

#[test]
fn can_clear_ack() {
    let mut dev = new(&[I2cTrans::read(Address::ARA, vec![0])]);
    dev.clear_ack().unwrap();
    destroy(dev);
}

#[test]
fn can_init() {
    let mut dev = new(&[
        I2cTrans::read(Address::ARA, vec![0]),
        I2cTrans::write(Address::COMMAND, vec![DEFAULT_CMD]),
    ]);
    dev.init().unwrap();
    destroy(dev);
}

#[test]
fn can_enable() {
    let mut dev = new(&[I2cTrans::write(Address::COMMAND, vec![DEFAULT_CMD])]);
    dev.enable().unwrap();
    destroy(dev);
}

#[test]
fn can_disable() {
    let mut dev = new(&[I2cTrans::write(
        Address::COMMAND,
        vec![DEFAULT_CMD | BitFlags::SHUTDOWN],
    )]);
    dev.disable().unwrap();
    destroy(dev);
}

#[test]
fn can_read_uv() {
    let mut dev = new(&[
        I2cTrans::read(Address::DATA_MSB, vec![0xAB]),
        I2cTrans::read(Address::DATA_LSB, vec![0xCD]),
    ]);
    let reading = dev.read_uv().unwrap();
    assert_eq!(0xABCD, reading);
    destroy(dev);
}

macro_rules! it_test {
    ( $test_name:ident, $it:expr, $expected:expr ) => {
        #[test]
        fn $test_name() {
            let mut dev = new(&[I2cTrans::write(
                Address::COMMAND,
                vec![DEFAULT_CMD | $expected],
            )]);
            dev.set_integration_time($it).unwrap();
            destroy(dev);
        }
    };
}

it_test!(can_set_integration_time_half_t, IntegrationTime::HalfT, 0);
it_test!(
    can_set_integration_time_1_t,
    IntegrationTime::T1,
    BitFlags::IT0
);
it_test!(
    can_set_integration_time_2_t,
    IntegrationTime::T2,
    BitFlags::IT1
);
it_test!(
    can_set_integration_time_4_t,
    IntegrationTime::T4,
    BitFlags::IT1 | BitFlags::IT0
);

#[test]
fn can_enable_ack() {
    let mut dev = new(&[I2cTrans::write(
        Address::COMMAND,
        vec![DEFAULT_CMD | BitFlags::ACK],
    )]);
    dev.enable_ack().unwrap();
    destroy(dev);
}

#[test]
fn can_disable_ack() {
    let mut dev = new(&[I2cTrans::write(Address::COMMAND, vec![DEFAULT_CMD])]);
    dev.disable_ack().unwrap();
    destroy(dev);
}

#[test]
fn can_set_ack_threshold_102_steps() {
    let mut dev = new(&[I2cTrans::write(Address::COMMAND, vec![DEFAULT_CMD])]);
    dev.set_ack_threshold(AckThreshold::Steps102).unwrap();
    destroy(dev);
}

#[test]
fn can_set_ack_threshold_145_steps() {
    let mut dev = new(&[I2cTrans::write(
        Address::COMMAND,
        vec![DEFAULT_CMD | BitFlags::ACK_THD],
    )]);
    dev.set_ack_threshold(AckThreshold::Steps145).unwrap();
    destroy(dev);
}

#[test]
fn can_enable_ack_with_threshold_145_steps() {
    let mut dev = new(&[I2cTrans::write(
        Address::COMMAND,
        vec![DEFAULT_CMD | BitFlags::ACK | BitFlags::ACK_THD],
    )]);
    dev.enable_ack_with_threshold(AckThreshold::Steps145)
        .unwrap();
    destroy(dev);
}


src/lib.rs
//! This is a platform agnostic Rust driver for the VEML6070 UVA light
//! sensor, based on the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Enable/disable the sensor. See: [`enable()`].
//! - Read the UV measurement. See: [`read_uv()`].
//! - Set the integration time. See: [`set_integration_time()`].
//! - Enable/disable ACK signal. See: [`enable_ack()`].
//! - Set ACK threshold value. See: [`set_ack_threshold()`].
//!
//! [`enable()`]: struct.Veml6070.html#method.enable
//! [`read_uv()`]: struct.Veml6070.html#method.read_uv
//! [`set_integration_time()`]: struct.Veml6070.html#method.set_integration_time
//! [`enable_ack()`]: struct.Veml6070.html#method.enable_ack
//! [`set_ack_threshold()`]: struct.Veml6070.html#method.set_ack_threshold
//!
//! ## The device
//! VEML6070 is an advanced ultraviolet (UV) light sensor with I2C protocol
//! interface and designed by the CMOS process.
//! It is easily operated via a simple I2C command. The active acknowledge
//! (ACK) feature with threshold windows setting allows the UV sensor to
//! send out a UVI alert message. Under a strong solar UVI condition, the
//! smart ACK signal can be easily implemented by the software programming.
//! VEML6070 incorporates a photodiode, amplifiers, and analog / digital
//! circuits into a single chip. VEML6070's adoption of Filtron TM UV
//! technology provides the best spectral sensitivity to cover UV spectrum
//! sensing. It has an excellent temperature compensation and a robust refresh
//! rate setting that does not use an external RC low pass filter.
//! VEML6070 has linear sensitivity to solar UV light and is easily adjusted
//! by an external resistor.
//! Software shutdown mode is provided, which reduces power consumption to
//! be less than 1 μA. VEML6070's operating voltage ranges from 2.7 V to 5.5 V.
//!
//! Datasheet:
//! - [VEML6070](https://www.vishay.com/docs/84277/veml6070.pdf)
//!
//! Application note:
//! - [VEML6070 AN](https://www.vishay.com/docs/84310/designingveml6070.pdf)
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then instantiate the device.
//!
//! Please find additional examples in this repository: [driver-examples]
//!
//! [driver-examples]: https://github.com/eldruin/driver-examples
//!
//! ### Read UV
//!
//! Import this crate and an `embedded_hal` implementation, then instantiate
//! the device:
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6070::Veml6070;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut uv_light_sensor = Veml6070::new(dev);
//! // initialization step is necessary
//! uv_light_sensor.init().unwrap();
//! uv_light_sensor.enable().unwrap();
//! let _uv_reading = uv_light_sensor.read_uv().unwrap();
//! ```
//!
//! ### Set integration time
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6070::{Veml6070, IntegrationTime};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut uv_light_sensor = Veml6070::new(dev);
//! // initialization step is necessary
//! uv_light_sensor.init().unwrap();
//! uv_light_sensor.enable().unwrap();
//! uv_light_sensor.set_integration_time(IntegrationTime::T1).unwrap();
//! ```
//!
//! ### Enable ACK and set a threshold of 145 steps
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use veml6070::{Veml6070, AckThreshold};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut uv_light_sensor = Veml6070::new(dev);
//! // initialization step is necessary
//! uv_light_sensor.init().unwrap();
//! uv_light_sensor.enable().unwrap();
//! uv_light_sensor.enable_ack_with_threshold(AckThreshold::Steps145).unwrap();
//! ```

#![doc(html_root_url = "https://docs.rs/veml6070/0.1.0")]
#![deny(unsafe_code)]
#![deny(missing_docs)]
#![no_std]

use embedded_hal::blocking::i2c;

/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C bus error
    I2C(E),
}

/// Integration time
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum IntegrationTime {
    /// Half T
    HalfT,
    /// 1 T
    T1,
    /// 2 T
    T2,
    /// 4 T
    T4,
}

/// ACK threshold
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AckThreshold {
    /// 102 steps
    Steps102,
    /// 145 steps
    Steps145,
}

struct BitFlags;

impl BitFlags {
    const SHUTDOWN: u8 = 0b0000_0001;
    const IT0: u8 = 0b0000_0100;
    const IT1: u8 = 0b0000_1000;
    const ACK_THD: u8 = 0b0001_0000;
    const ACK: u8 = 0b0010_0000;
}

struct Address;

impl Address {
    const ARA: u8 = 0x0C;
    const COMMAND: u8 = 0x38;
    const DATA_MSB: u8 = 0x39;
    const DATA_LSB: u8 = 0x38;
}

const DEFAULT_CMD: u8 = 0x02;

/// VEML6070 device driver.
#[derive(Debug)]
pub struct Veml6070<I2C> {
    /// The concrete I²C device implementation.
    i2c: I2C,
    /// Command register status.
    cmd: u8,
}

impl<I2C, E> Veml6070<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    /// Create new instance of the VEML6070 device.
    pub fn new(i2c: I2C) -> Self {
        Veml6070 {
            i2c,
            cmd: DEFAULT_CMD,
        }
    }

    /// Destroy driver instance, return I²C bus instance.
    pub fn destroy(self) -> I2C {
        self.i2c
    }

    /// Enable the sensor.
    pub fn enable(&mut self) -> Result<(), Error<E>> {
        let cmd = self.cmd;
        self.write_command(cmd & !BitFlags::SHUTDOWN)
    }

    /// Disable the sensor (shutdown).
    pub fn disable(&mut self) -> Result<(), Error<E>> {
        let cmd = self.cmd;
        self.write_command(cmd | BitFlags::SHUTDOWN)
    }

    /// Set integration time.
    pub fn set_integration_time(&mut self, it: IntegrationTime) -> Result<(), Error<E>> {
        let mut cmd = self.cmd;
        cmd = match it {
            IntegrationTime::HalfT => cmd & !BitFlags::IT0 & !BitFlags::IT1,
            IntegrationTime::T1 => cmd | BitFlags::IT0 & !BitFlags::IT1,
            IntegrationTime::T2 => cmd & !BitFlags::IT0 | BitFlags::IT1,
            IntegrationTime::T4 => cmd | BitFlags::IT0 | BitFlags::IT1,
        };
        self.write_command(cmd)
    }

    /// Enable the ACK signal.
    ///
    /// *Note:* The ACK must be cleared every time after it has fired with `clear_ack()`.
    pub fn enable_ack(&mut self) -> Result<(), Error<E>> {
        let cmd = self.cmd;
        self.write_command(cmd | BitFlags::ACK)
    }

    /// Disable the ACK signal.
    pub fn disable_ack(&mut self) -> Result<(), Error<E>> {
        let cmd = self.cmd;
        self.write_command(cmd & !BitFlags::ACK)
    }

    /// Set ACK threshold.
    pub fn set_ack_threshold(&mut self, threshold: AckThreshold) -> Result<(), Error<E>> {
        let cmd = self.cmd;
        self.write_command(handle_ack_threshold_bit(cmd, threshold))
    }

    /// Enable the ACK signal and set the ACK threshold at once.
    ///
    /// *Note:* The ACK must be cleared every time after it has fired with `clear_ack()`.
    pub fn enable_ack_with_threshold(&mut self, threshold: AckThreshold) -> Result<(), Error<E>> {
        let mut cmd = self.cmd;
        cmd |= BitFlags::ACK;
        self.write_command(handle_ack_threshold_bit(cmd, threshold))
    }

    fn write_command(&mut self, cmd: u8) -> Result<(), Error<E>> {
        self.i2c
            .write(Address::COMMAND, &[cmd])
            .map_err(Error::I2C)?;
        self.cmd = cmd;
        Ok(())
    }
}

fn handle_ack_threshold_bit(cmd: u8, threshold: AckThreshold) -> u8 {
    match threshold {
        AckThreshold::Steps102 => cmd & !BitFlags::ACK_THD,
        AckThreshold::Steps145 => cmd | BitFlags::ACK_THD,
    }
}

impl<I2C, E> Veml6070<I2C>
where
    I2C: i2c::Read<Error = E>,
{
    /// Clear ACK status.
    ///
    /// *Note:* The ACK status must be cleared every time after it has fired.
    /// Other registers will be blocked otherwise.
    pub fn clear_ack(&mut self) -> Result<(), Error<E>> {
        let mut buffer = [0];
        self.i2c.read(Address::ARA, &mut buffer).map_err(Error::I2C)
    }

    /// Read the UV sensor.
    pub fn read_uv(&mut self) -> Result<u16, Error<E>> {
        let mut msb = [0];
        let mut lsb = [0];
        self.i2c
            .read(Address::DATA_MSB, &mut msb)
            .map_err(Error::I2C)?;
        self.i2c
            .read(Address::DATA_LSB, &mut lsb)
            .map_err(Error::I2C)?;
        Ok(u16::from(msb[0]) << 8 | u16::from(lsb[0]))
    }
}

impl<I2C, E> Veml6070<I2C>
where
    I2C: i2c::Read<Error = E> + i2c::Write<Error = E>,
{
    /// Initialize and clear ACK.
    pub fn init(&mut self) -> Result<(), Error<E>> {
        self.clear_ack()?;
        self.write_command(DEFAULT_CMD)
    }
}
